rpc/v8/block.go

```diff
@@ -3,56 +3,125 @@ package rpcv8
 import (
 	"encoding/json"
 	"errors"
+	"fmt"
 
 	"github.com/NethermindEth/juno/core"
 	"github.com/NethermindEth/juno/core/felt"
 	"github.com/NethermindEth/juno/jsonrpc"
 	rpcv6 "github.com/NethermindEth/juno/rpc/v6"
 )
 
+var _ BlockIdentifier = (*BlockID)(nil)
+
+type BlockIdentifier interface {
+	Type() blockIDType
+	IsLatest() bool
+	IsPending() bool
+	IsHash() bool
+	IsNumber() bool
+	Hash() *felt.Felt
+	Number() uint64
+	UnmarshalJSON(data []byte) error
+}
+
+type blockIDType uint8
+
+const (
+	pending blockIDType = iota + 1
+	latest
+	hash
+	number
+)
+
+func (b *blockIDType) String() string {
+	switch *b {
+	case pending:
+		return "pending"
+	case latest:
+		return "latest"
+	case hash:
+		return "hash"
+	case number:
+		return "number"
+	default:
+		panic(fmt.Sprintf("Unknown blockIdType: %d", b))
+	}
+}
+
 // https://github.com/starkware-libs/starknet-specs/blob/a789ccc3432c57777beceaa53a34a7ae2f25fda0/api/starknet_api_openrpc.json#L814
 type BlockID struct {
-	Pending bool
-	Latest  bool
-	Hash    *felt.Felt
-	Number  uint64
+	typeID blockIDType
```

Comment:
Why don't we have 4 types implementing `BlockIdentifier`? Something like
```
type PendingBlock struct{}
type LatestBlock struct{}
type BlockByHash felt.Felt
type BlockByNumber uint64
```

Reply:
Some rationale:
- We don't allocate `data` if not necessary.
- `Hash()` or `Number()` are unsafe functions, so we need to make sure that we only call them in correct cases. If we can instead extract them directly from the switch cases, I think it's safer. Something like:
```
switch blockId := blockId.(type) {
case BlockByHash:
  hash := felt.Felt(blockId)
...
```

Reply:
If we keep the current `BlockID` implementation, I think we can remove `BlockIdentifier`, because there is only 1 implementation and we don't hide it anyway.